File: 53.最大子序和.py
#
# @lc app=leetcode.cn id=53 lang=python3
#
# [53] 最大子序和
#
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) == 0: 
            return 0
        if len(nums) == 1: 
            return nums[0]
        
        maxSolution: int = nums[0]
        curMaxSolution: int = nums[0]
        
        for idx in range(1, len(nums)):
            curMaxSolution = max(curMaxSolution + nums[idx], nums[idx])
            maxSolution = max(curMaxSolution, maxSolution)
        
        return maxSolution

File: 53.最大子序和.java
/*
 * @lc app=leetcode.cn id=53 lang=java
 *
 * [53] 最大子序和
 */
class Solution {
    public int maxSubArray(int[] nums) {
        if (nums.length == 0) return 0;
        if (nums.length == 1) return nums[0];
        int curMaxSolution = nums[0];
        int maxSolution = curMaxSolution;
        for (int i = 1; i < nums.length; i++) {
            curMaxSolution = curMaxSolution <= 0 ? nums[i] : curMaxSolution + nums[i];
            maxSolution = Math.max(curMaxSolution, maxSolution);
        }
        return maxSolution;
    }
}

